libs/providers/multi-provider/src/lib/strategies/BaseEvaluationStrategy.ts

```diff
@@ -0,0 +1,92 @@
+import {
+  EvaluationContext,
+  FlagValue,
+  FlagValueType,
+  Provider,
+  ProviderStatus,
+  ResolutionDetails,
+} from '@openfeature/server-sdk';
+import { ErrorWithCode } from '@openfeature/multi-provider';
+
+export type StrategyEvaluationContext = {
+  flagKey: string;
+  flagType: FlagValueType;
+};
+export type StrategyPerProviderContext = StrategyEvaluationContext & {
+  provider: Provider;
+  providerName: string;
+  providerStatus: ProviderStatus;
+};
+
+export type ProviderResolutionResult<T extends FlagValue> = {
+  provider: Provider;
+  providerName: string;
+  thrownError?: unknown;
+  details?: ResolutionDetails<T>;
+};
```

Comment:
Instead of `thrownError` and `details` being optional, consider something like the following (Not sure if the names are exactly right)

````ts
export type ProviderResolutionBaseResult = {
  provider: Provider;
  providerName: string;
};

export type ProviderResolutionSuccessResult<T extends FlagValue> = ProviderResolutionBaseResult & {
  details: ResolutionDetails<T>;
};

export type ProviderResolutionErrorResult = ProviderResolutionBaseResult & {
  thrownError: unknown;
};

export type ProviderResolutionResult<T extends FlagValue> =
  | ProviderResolutionErrorResult
  | ProviderResolutionSuccessResult<T>;
````

This way we can avoid type assertions like here

Reply:
For me it worked 🤔

Reply:
got it working, the only caveat is that I couldn't get the types of the `hasError` check in `determineResult` to narrow well enough, and ended up moving that logic inline

Reply:
Cool!
Well it is possible, if you write the `hasError` function this way:

```ts
  protected hasError(resolution: ProviderResolutionResult<FlagValue>): resolution is
    | ProviderResolutionErrorResult
    | (ProviderResolutionSuccessResult<FlagValue> & {
        details: ResolutionDetails<FlagValue> & { errorCode: ErrorCode };
      }) {
    return 'thrownError' in resolution || !!resolution.details.errorCode;
  }
```

This works, tested it, but the type is ugly :D Sure one could define that type where `ProviderResolutionResult` is defined, but maybe it is just fine to do it inline. I am not sure.